### ISSUE TITLE
New lint that warns on all structs and enums not marked #[non_exhaustive]

### ISSUE THREAD

**Issue by kangalio** [rust-lang/rust-clippy#10687]
### What it does

Emit a warning for every struct and enum that's not marked #[non_exhaustive]

### Lint Name

missing_non_exhaustive

### Category

restriction

### Advantage

Some codebases need to add fields to structs/enums regularly (for example if they model an upstream JSON schema). Those structs/enums should almost all be marked #[non_exhaustive], or else you cannot incorporate new fields without a breaking release.

In [serenity](https://github.com/serenity-rs/serenity), we would apply `#[warn(clippy::missing_non_exhaustive)]` to the entire [`serenity::model`](https://docs.rs/serenity/latest/serenity/model/index.html) submodule. The few odd structs that _should_ be exhaustive would be marked `#[allow(clippy::missing_non_exhaustive)]`

### Drawbacks

None, because this is opt-in

### Example

```rust
pub struct Foo {
    pub bar: String,
}
```

Could be written as:

```rust
#[non_exhaustive]
pub struct Foo {
    pub bar: String,
}
```

**Comment by kangalio** [comment]
You're right, I don't know how I missed that
